core/src/main/scala/org/apache/spark/scheduler/DecommissionTracker.scala

```diff
@@ -0,0 +1,405 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.scheduler
+
+import java.text.SimpleDateFormat
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicInteger}
+
+import scala.collection.mutable.HashMap
+
+import org.apache.spark.{ExecutorAllocationClient, SparkConf, SparkContext}
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config
+import org.apache.spark.util.{Clock, SystemClock, ThreadUtils, Utils}
+
+/**
+ * DecommissionTracker tracks the list of decommissioned nodes.
+ * This decommission trackers maintains the decommissioned nodes state.
+ * Decommission tracker schedules the executor decommission and shuffle
+ * decommission for that node.
+ */
+private[scheduler] class DecommissionTracker (
+  conf: SparkConf,
+  executorAllocClient: Option[ExecutorAllocationClient],
+  dagScheduler: Option[DAGScheduler],
+  clock: Clock = new SystemClock()) extends Logging {
+
+  def this(sc: SparkContext,
+           client: Option[ExecutorAllocationClient],
+           dagScheduler: Option[DAGScheduler]) = {
+    this(sc.conf, client, dagScheduler)
+  }
+
+  // Decommission thread of node decommissioning!!
+  private val decommissionThread =
+    ThreadUtils.newDaemonThreadPoolScheduledExecutor("node-decommissioning-thread", 20)
```

Comment:
We have created a threadpool for 20 here, For handling the scenario when there are lots of node decommissioning happened at same time. But we can add this as part of config . Shall I made this as a part of config?

Reply:
20 threads to just handle node decommissions seems like a lot.  How long does it normally take for a decommission? On Yarn I realize it has to send message to AM which could then be blocked on doing allocate call or something, so I'm assuming that is where possible delay is.  But have you actually run this and decommissioned multiple nodes, how long does it take? 
You are also calling executedecomission separate from the shuffle decommission and then have delays to have them happen not at the same time. Couldn't they be done in sequence and then you know they aren't done at same time?  
what was the default yarn timeout on newer versions?  I know you said 30 seconds on gap.

Reply:
So the above code creates the thread pool of 20. A new thread will be created if none of the thread is free in the thread pool after scheduled time for the executors and shuffle decommission(executorsDecomissionTimeMs and ShuffleDecosmmissionTimeMs). These active threads completes the task in less than second (for executorDecommission it will kill all the executors on the host and for shuffleDecommission remove all the entries of shuffle data for that node from the MapOutputTracker).

**You are also calling executedecomission separate from the shuffle decommission and then have delays to have them happen not at the same time. Couldn't they be done in sequence and then you know they aren't done at same time?** - So the reason why it is not done in sequence . If inside  executordecommission after decommission of executors the same thread wait for some time and than do the shuffleDecommission . In this case we are holding thread for long time which can be free in 1 sec. Now consider a scenario where aws spotloss happened after 120 sec of receiving the decommission event, As per the logic written executor decommission will takes place after 50% time i.e. after 60 sec  and shuffleDecommission will happen after 90% of the time 108 sec. If we do this sequentially than we have to hold current thread for another (108-60=48) 48 secs. So there are more chances of creating more thread from the pool if we do it sequentially.

Now as per the current code the executor decommission will takes place after 50% time i.e. after 60 sec. ExecutorDecommissionThread can take one of the thread completes its task in max 1 sec and than that thread is free and same is for shuffle decommission which will happened after 90% time 108sec completes the task in 1 sec and than the thread is free again. This free thread can be used by another Executordecommision/Shuffle decommission if decommission is received for  multiple nodes.

Its very rare scenario when all the 20 threads would be active where the decommission is very frequent. I think we can have the config for the thread pool and believe 8 would be good number and if someone facing the issue very frequent  decommission than they can change it